### PR TITLE
Rewrite [pypi]; affects [npm]

### PR DESCRIPTION
**Pull request by paulmelnikow**
When I worked on NPM and UptimeRobot, I created separate PRs for the refactors. Does that approach make sense here? Or would it be just as easy to review this in total?

Todo:

- [x] Fix broken tests
- [x] Contract
- [ ] Tests
    - [ ] Deprecated()
    - [ ] enforceDeprecation()
    - [ ] renderVersionBadge()
    - [x] getPackageFormats()

Ref: #1358

### CODE REVIEW

File: services/npm/npm-license.service.js [hunk 1]
```diff
@@ -9,10 +9,6 @@ module.exports = class NpmLicense extends NpmBase {
     return 'license'
   }
 
-  static get defaultBadgeData() {
-    return { label: 'license' }
-  }
-
```
Comment (paulmelnikow) [inline services/npm/npm-license.service.js]: This already defaults to `this.category` which is `'license'`.

File: services/npm/npm.tester.js [hunk 2]
```diff
@@ -213,7 +213,6 @@ t.create('license when network is off')
 // This tests error-handling functionality in NpmBase.
 t.create('when json is malformed for scoped package')
   .get('/l/@cycle%2Fcore.json')
-  .only()
```
Comment (chris48s) [inline services/npm/npm.tester.js]: I haven't looked at the rest of this, but can we get this change in ahead of the rest of this PR. We've got other open PRS ( #1897 #1740 ) which should run against all the NPM tests.

Also, there's a couple of times a `.only()` has been missed in review. Maybe we should have a `danger` rule for this..

Comment (paulmelnikow) [reply services/npm/npm.tester.js]: Done in #1926.

File: lib/deprecation-helpers.js [hunk 3]
```diff
@@ -21,7 +22,14 @@ const getDeprecatedBadge = function(label, data) {
   return badgeData
 }
 
+function enforceDeprecation(effectiveDate) {
+  if (Date.now() >= effectiveDate.getTime()) {
```
Comment (RedSparr0w) [inline lib/deprecation-helpers.js]: `Date.now()`
Never knew this was a thing 👍

**Comment by paulmelnikow** [comment]
Echoing my question from the top post…

> When I worked on NPM and UptimeRobot, I created separate PRs for the refactors. Does that approach make sense here? Or would it be just as easy to review this in total?

Any thoughts?

If this needs to be split up, I want to do that before/as I clean it up, not after I clean it up.

**Comment by chris48s** [comment]
_In theory_ it is nicer to review stuff in smaller chunks, but given the original implementation defines a bunch of badges in one huge function I reckon trying to do that in this case will introduce more intermediate complexity than its worth.

**Comment by paulmelnikow** [comment]
Okay. Agreed the pypi badges should just come in all at once. I was thinking I _could_ separate out the changes in the base functionality, and the little bit of refactoring in npm. Let me know if you think it’s worth it.

File: services/pypi/pypi-djversions.service.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+'use strict'
+
+const PypiBase = require('./pypi-base')
+const { sortDjangoVersions, parseClassifiers } = require('./pypi-helpers')
+
+module.exports = class PypiDjangoVersions extends PypiBase {
+  static get category() {
+    return 'version'
```
Comment (PyvesB) [inline services/pypi/pypi-djversions.service.js]: This should be updated to 'platform-support' following #1948.

File: services/pypi/pypi-pyversions.service.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
+'use strict'
+
+const PypiBase = require('./pypi-base')
+const { parseClassifiers } = require('./pypi-helpers')
+
+module.exports = class PypiPythonVersions extends PypiBase {
+  static get category() {
+    return 'version'
```
Comment (PyvesB) [inline services/pypi/pypi-pyversions.service.js]: This should be updated to 'platform-support' following #1948.

File: services/pypi/pypi-pyversions.service.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        keywords: ['python'],
+      },
+    ]
+  }
+
+  static render({ versions }) {
+    const versionSet = new Set(versions)
+    // We only show v2 if eg. v2.4 does not appear.
+    // See https://github.com/badges/shields/pull/489 for more.
+    ;['2', '3'].forEach(majorVersion => {
```
Comment (PyvesB) [inline services/pypi/pypi-pyversions.service.js]: What is this semicolon doing here? 😄

File: services/pypi/pypi-helpers.js [hunk 7]
```diff
@@ -50,8 +50,25 @@ const parseClassifiers = function(parsedData, pattern) {
   return results
 }
 
+function getPackageFormats(packageData) {
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply services/pypi/pypi-helpers.js]: Good idea!

[the remark this answers is not in the document]
Comment (paulmelnikow) [reply services/pypi/pypi-helpers.js]: Updated.

File: services/pypi/pypi-pyversions.service.js (hunk 6 above)
Comment (PyvesB) [reply services/pypi/pypi-pyversions.service.js]: This makes much more sense now. Thanks for explaining!

**Comment by paulmelnikow** [comment]
Thanks! I will get this merged and then circle back on the few remaining tests.

**Pull request merged**